### PR TITLE
[SPARK-40826][SS] Add additional checkpoint rename file check

### PR DESCRIPTION
**Pull request by viirya**
### What changes were proposed in this pull request?

This adds additional checkpoint rename file check.

### Why are the changes needed?

We encountered an issue recently that one customer's structured streaming job failed to read delta file.

The temporary file exists but it was not successfully renamed to final delta file path.

We currently don't check if renamed file exists but assume it successful. As the result, failing to read delta file assumed to be committed in last batch makes re-triggering the job impossible.

We should be able to do a check against checkpoint renamed file to prevent such difficulty in advance.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Existing tests.

### CODE REVIEW

**Comment by viirya** [comment]
cc @dongjoon-hyun

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 1]
```diff
@@ -1830,6 +1830,13 @@ object SQLConf {
       .stringConf
       .createWithDefault("lz4")
 
+  val CHECKPOINT_RENAMEDFILE_CHECK_ENABLED =
+    buildConf("spark.sql.streaming.checkpoint.renamedFileCheck")
```
Comment (itholic) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:1834]: I think maybe we usually "enabled" keyword for string when "ENABLED" is following in the postfix for option?

So, for example:
```diff
-  "spark.sql.streaming.checkpoint.renamedFileCheck"
+  "spark.sql.streaming.checkpoint.renamedFileCheck.enabled"
```

WDYT?

**Review by itholic — COMMENTED** [review]
Looks pretty good otherwise

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 1 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:1834]: Okay. LGTM.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 2]
```diff
@@ -1830,6 +1830,13 @@ object SQLConf {
       .stringConf
       .createWithDefault("lz4")
 
+  val CHECKPOINT_RENAMEDFILE_CHECK_ENABLED =
+    buildConf("spark.sql.streaming.checkpoint.renamedFileCheck.enabled")
+      .doc("When true, Spark will validate if renamed checkpoint file exists.")
+      .internal()
+      .booleanConf
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:1837]:
```
.version("3.4.0")
```

**Review by dongjoon-hyun — APPROVED** [review]

**Comment by dongjoon-hyun** [comment]
Are the failures related?
```
[info] *** 12 TESTS FAILED ***
[error] Failed: Total 6746, Failed 12, Errors 0, Passed 6734, Ignored 5
[error] Failed tests:
[error] 	org.apache.spark.sql.catalyst.expressions.CastWithAnsiOffSuite
[error] 	org.apache.spark.sql.catalyst.util.TimestampFormatterSuite
[error] 	org.apache.spark.sql.catalyst.expressions.CastWithAnsiOnSuite
[error] 	org.apache.spark.sql.catalyst.util.RebaseDateTimeSuite
[error] 	org.apache.spark.sql.catalyst.expressions.TryCastSuite
```

**Comment by dongjoon-hyun** [comment]
Since the first commit passed the UT, I guess it's irrelevant.

**Comment by viirya** [comment]
It is not. I only add `version` in the second commit.

**Comment by viirya** [comment]
Verified locally these tests are passed.

**Comment by viirya** [comment]
Synced up with master and try again.

**Review by HyukjinKwon — APPROVED** [review]

**Comment by dongjoon-hyun** [comment]
Ya, `master` branch seems to be broken with the same irrelevant failures, too.
- https://github.com/apache/spark/actions/runs/3278495669/jobs/5397028188

**Comment by dongjoon-hyun** [comment]
Merged to master for Apache Spark 3.4.
Thank you, @viirya , @itholic , @HyukjinKwon .

**Comment by viirya** [comment]
Thanks @dongjoon-hyun @HyukjinKwon @itholic

**Comment by HeartSaVioR** [comment]
(Just curious about how you came through the case. Was the issue from underlying storage, or inconsistency of file system API? I can't imagine the scenario if the API works as contract and the underlying file system is stable, so...)

**Pull request closed without merging**